### ISSUE TITLE
Problem with python-cffi while installing Klipper

### ISSUE THREAD

**Issue by ntx007** [KevinOConnor/klipper#2480]
(no Klipper log existing)

Hello,

i have a problem with installing klipper on my octopi. 
When i start/execute "./klipper/scripts/install-octopi.sh" i get a error from "cffi". 
I have looked on multiple internet sites but without any luck. Also i tried to install cffi manually but still with errors.

Here is the command line history:

> `pi@octopi:~ $ ./klipper/scripts/install-octopi.sh
> 
> 
> ###### Running apt-get update...
> [sudo] password for pi: 
> Hit:1 http://raspbian.raspberrypi.org/raspbian buster InRelease
> Hit:2 http://archive.raspberrypi.org/debian buster InRelease
> Reading package lists... Done
> 
> 
> ###### Installing packages...
> Reading package lists... Done
> Building dependency tree       
> Reading state information... Done
> avr-libc is already the newest version (1:2.0.0+Atmel3.6.1-2).
> binutils-arm-none-eabi is already the newest version (2.31.1-11+rpi1+11).
> binutils-avr is already the newest version (2.26.20160125+Atmel3.6.1-4).
> build-essential is already the newest version (12.6).
> dfu-util is already the newest version (0.9-1).
> gcc-arm-none-eabi is already the newest version (15:7-2018-q2-6).
> gcc-avr is already the newest version (1:5.4.0+Atmel3.6.1-2).
> libffi-dev is already the newest version (3.2.1-9).
> libncurses-dev is already the newest version (6.1+20181013-2+deb10u2).
> libnewlib-arm-none-eabi is already the newest version (3.1.0.20181231-1).
> libusb-dev is already the newest version (2:0.1.12-32).
> python-dev is already the newest version (2.7.16-1).
> python-virtualenv is already the newest version (15.1.0+ds-2).
> stm32flash is already the newest version (0.5-1).
> virtualenv is already the newest version (15.1.0+ds-2).
> avrdude is already the newest version (6.3-20171130+svn1429-2+rpt1).
> 0 upgraded, 0 newly installed, 0 to remove and 2 not upgraded.
> 
> 
> ###### Updating python virtual environment...
> DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
> Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
> Collecting cffi==1.12.2
>   Using cached cffi-1.12.2.tar.gz (453 kB)
> Collecting pyserial==3.4
>   Using cached pyserial-3.4-py2.py3-none-any.whl (193 kB)
> Collecting greenlet==0.4.15
>   Using cached greenlet-0.4.15.tar.gz (59 kB)
> Collecting Jinja2==2.10.1
>   Using cached Jinja2-2.10.1-py2.py3-none-any.whl (124 kB)
> Requirement already satisfied: pycparser in ./klippy-env/lib/python2.7/site-packages (from cffi==1.12.2->-r /home/pi/klipper/scripts/klippy-requirements.txt (line 5)) (2.19)
> Processing ./.cache/pip/wheels/f2/aa/04/0edf07a1b8a5f5f1aed7580fffb69ce8972edc16a505916a77/MarkupSafe-1.1.1-cp27-cp27mu-linux_armv7l.whl
> Building wheels for collected packages: cffi, greenlet
>   Building wheel for cffi (setup.py) ... error
>   ERROR: Command errored out with exit status 1:
>    command: /home/pi/klippy-env/bin/python2 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-l3DVIq/cffi/setup.py'"'"'; __file__='"'"'/tmp/pip-install-l3DVIq/cffi/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-u9GEbb
>        cwd: /tmp/pip-install-l3DVIq/cffi/
>   Complete output (46 lines):
>   running bdist_wheel
>   running build
>   running build_py
>   creating build
>   creating build/lib.linux-armv7l-2.7
>   creating build/lib.linux-armv7l-2.7/cffi
>   copying cffi/commontypes.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/pkgconfig.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/api.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/ffiplatform.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/recompiler.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/setuptools_ext.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/model.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/vengine_gen.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/vengine_cpy.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/backend_ctypes.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/__init__.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/lock.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/verifier.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/cparser.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/error.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/cffi_opcode.py -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/_cffi_include.h -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/parse_c_type.h -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/_embedding.h -> build/lib.linux-armv7l-2.7/cffi
>   copying cffi/_cffi_errors.h -> build/lib.linux-armv7l-2.7/cffi
>   running build_ext
>   building '_cffi_backend' extension
>   creating build/temp.linux-armv7l-2.7
>   creating build/temp.linux-armv7l-2.7/c
>   arm-linux-gnueabihf-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fdebug-prefix-map=/build/python2.7-InigCj/python2.7-2.7.16=. -fstack-protector-strong -Wformat -Werror=format-security -fPIC -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/usr/include/python2.7 -c c/_cffi_backend.c -o build/temp.linux-armv7l-2.7/c/_cffi_backend.o
>   during GIMPLE pass: pre
>   In file included from c/_cffi_backend.c:323:
>   c/minibuffer.h: In function ‘mb_clear’:
>   c/minibuffer.h:157:1: internal compiler error: Segmentation fault
>    mb_clear(MiniBufferObj *ob)
>    ^~~~~~~~
>   0x76b5211f ???
>   	../sysdeps/unix/sysv/linux/arm/sigrestorer.S:64
>   0x76b3c717 __libc_start_main
>   	/build/glibc-FUvrFr/glibc-2.28/csu/libc-start.c:308
>   Please submit a full bug report,
>   with preprocessed source if appropriate.
>   Please include the complete backtrace with any bug report.
>   See <file:///usr/share/doc/gcc-8/README.Bugs> for instructions.
>   error: command 'arm-linux-gnueabihf-gcc' failed with exit status 1
>   ----------------------------------------
>   ERROR: Failed building wheel for cffi
>   Running setup.py clean for cffi
>   Building wheel for greenlet (setup.py) ... error
>   ERROR: Command errored out with exit status 1:
>    command: /home/pi/klippy-env/bin/python2 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-l3DVIq/greenlet/setup.py'"'"'; __file__='"'"'/tmp/pip-install-l3DVIq/greenlet/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-6zL47z
>        cwd: /tmp/pip-install-l3DVIq/greenlet/
>   Complete output (21 lines):
>   running bdist_wheel
>   running build
>   running build_ext
>   building 'greenlet' extension
>   creating build
>   creating build/temp.linux-armv7l-2.7
>   arm-linux-gnueabihf-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fdebug-prefix-map=/build/python2.7-InigCj/python2.7-2.7.16=. -fstack-protector-strong -Wformat -Werror=format-security -fPIC -I/usr/include/python2.7 -c greenlet.c -o build/temp.linux-armv7l-2.7/greenlet.o
>   during GIMPLE pass: pre
>   greenlet.c: In function ‘green_is_gc’:
>   greenlet.c:964:12: internal compiler error: Segmentation fault
>    static int green_is_gc(PyGreenlet* self)
>               ^~~~~~~~~~~
>   0x76b2c11f ???
>   	../sysdeps/unix/sysv/linux/arm/sigrestorer.S:64
>   0x76b16717 __libc_start_main
>   	/build/glibc-FUvrFr/glibc-2.28/csu/libc-start.c:308
>   Please submit a full bug report,
>   with preprocessed source if appropriate.
>   Please include the complete backtrace with any bug report.
>   See <file:///usr/share/doc/gcc-8/README.Bugs> for instructions.
>   error: command 'arm-linux-gnueabihf-gcc' failed with exit status 1
>   ----------------------------------------
>   ERROR: Failed building wheel for greenlet
>   Running setup.py clean for greenlet
> Failed to build cffi greenlet
> Installing collected packages: cffi, pyserial, greenlet, MarkupSafe, Jinja2
>     Running setup.py install for cffi ... error
>     ERROR: Command errored out with exit status 1:
>      command: /home/pi/klippy-env/bin/python2 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-l3DVIq/cffi/setup.py'"'"'; __file__='"'"'/tmp/pip-install-l3DVIq/cffi/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-pLzlqX/install-record.txt --single-version-externally-managed --compile --install-headers /home/pi/klippy-env/include/site/python2.7/cffi
>          cwd: /tmp/pip-install-l3DVIq/cffi/
>     Complete output (46 lines):
>     running install
>     running build
>     running build_py
>     creating build
>     creating build/lib.linux-armv7l-2.7
>     creating build/lib.linux-armv7l-2.7/cffi
>     copying cffi/commontypes.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/pkgconfig.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/api.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/ffiplatform.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/recompiler.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/setuptools_ext.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/model.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/vengine_gen.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/vengine_cpy.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/backend_ctypes.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/__init__.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/lock.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/verifier.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/cparser.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/error.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/cffi_opcode.py -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/_cffi_include.h -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/parse_c_type.h -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/_embedding.h -> build/lib.linux-armv7l-2.7/cffi
>     copying cffi/_cffi_errors.h -> build/lib.linux-armv7l-2.7/cffi
>     running build_ext
>     building '_cffi_backend' extension
>     creating build/temp.linux-armv7l-2.7
>     creating build/temp.linux-armv7l-2.7/c
>     arm-linux-gnueabihf-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fdebug-prefix-map=/build/python2.7-InigCj/python2.7-2.7.16=. -fstack-protector-strong -Wformat -Werror=format-security -fPIC -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/usr/include/python2.7 -c c/_cffi_backend.c -o build/temp.linux-armv7l-2.7/c/_cffi_backend.o
>     during GIMPLE pass: pre
>     In file included from c/_cffi_backend.c:323:
>     c/minibuffer.h: In function ‘mb_clear’:
>     c/minibuffer.h:157:1: internal compiler error: Segmentation fault
>      mb_clear(MiniBufferObj *ob)
>      ^~~~~~~~
>     0x76b6111f ???
>     	../sysdeps/unix/sysv/linux/arm/sigrestorer.S:64
>     0x76b4b717 __libc_start_main
>     	/build/glibc-FUvrFr/glibc-2.28/csu/libc-start.c:308
>     Please submit a full bug report,
>     with preprocessed source if appropriate.
>     Please include the complete backtrace with any bug report.
>     See <file:///usr/share/doc/gcc-8/README.Bugs> for instructions.
>     error: command 'arm-linux-gnueabihf-gcc' failed with exit status 1
>     ----------------------------------------
> ERROR: Command errored out with exit status 1: /home/pi/klippy-env/bin/python2 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-l3DVIq/cffi/setup.py'"'"'; __file__='"'"'/tmp/pip-install-l3DVIq/cffi/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-pLzlqX/install-record.txt --single-version-externally-managed --compile --install-headers /home/pi/klippy-env/include/site/python2.7/cffi Check the logs for full command output.`
> 

My OctoPrint Version: 1.3.12 (stable) running on RaspPi 3 Model B
OctoPi Version 0.17

Is there anyone here who can help me out? 
I am pretty new around here

**Comment by klipper-gitissuebot** [comment]
Hi @ntx007,

It did not look like there was a Klipper log file attached to this ticket. The log file has been engineered to answer common questions the Klipper developers have about the software and its environment (software version, hardware type, configuration, event timing, and hundreds of other questions).

Unfortunately, too many people have opened tickets without providing the log. That consumes developer time; time that would be better spent enhancing the software. If this ticket references an event that has occurred while running the software then the Klipper log must be attached to this ticket. Otherwise, this ticket will be automatically closed in a few days.

For information on obtaining the Klipper log file see: https://github.com/KevinOConnor/klipper/blob/master/docs/Contact.md

The log can still be attached to this ticket - just add a comment and attach the log to that comment.

Best regards,
~ Your friendly GitIssueBot

PS: I'm just an automated script, not a human being.

**Comment by ld3300** [comment]
One google search suggested this: `sudo apt install libffi-dev`

**Comment by KevinOConnor** [comment]
The most common cause of issues like this is a corrupt SD card in the Raspberry Pi.  You may want to try a fresh install of OctoPi.

-Kevin

**Comment by ntx007** [comment]
> One google search suggested this: `sudo apt install libffi-dev`

Its already installed, i just tested it.



> The most common cause of issues like this is a corrupt SD card in the Raspberry Pi. You may want to try a fresh install of OctoPi.
> 
> -Kevin

I just tried it but i got the next problem with flashing klipper.bin...

When i try to flash klipper.bin (renamend firmware.bin) the printer screen shows nothing (its on but empty). I have tried flashing via USB (which didnt work), so i tried flashing it manually via SD because i have a Ender 3 (SKR 1.3 XYZ@TMC2209 E@LV8729 and E3D Hemera DD) but without any luck. I attached the menuconfig screenshot and one from octoprint and more 

I noticed that the device name changed from marlin to klipper (/dev/serial/by-id/usb-Klipper_lpc1768_0F70FF0D018139AF4042075CC42000F5-if00)
[klippy.log](https://github.com/KevinOConnor/klipper/files/4165436/klippy.log)
<img width="648" alt="Bildschirmfoto 2020-02-06 um 14 46 18" src="https://user-images.githubusercontent.com/17531910/73943107-79246280-48f0-11ea-9f10-0f1d5b58aaff.png">
<img width="1116" alt="Bildschirmfoto 2020-02-06 um 14 47 25" src="https://user-images.githubusercontent.com/17531910/73943122-804b7080-48f0-11ea-8a95-ac2d4749c1c0.png">


Thanks for helping
